### DISCUSSION TITLE
/digging-deeper/cache.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#228] [Announcements]
# /digging-deeper/cache.html

[[toc]] 简介 Goravel 提供了可拓展的缓存模块，该模块可以使用 facades.Cache() 进行操作。框架默认自带 memory 驱动，如需其他驱动，请查看对应的独立扩展包： 驱动 地址 ----------- -------------- Redis https://github.com/goravel/redis | 配置 在 c...

https://www.goravel.dev/zh/digging-deeper/cache.html

**Comment by diaojinlong** [comment]
The following error occurs when using reids

``` invalid driver: , only support memory, custom ```

**Reply by hwbrzzl** [reply]
Yes, please look into https://www.goravel.dev/digging-deeper/cache.html#introduction

**Comment by Zstranslators** [comment]
Does the cache driver support caching groups?

**Reply by hwbrzzl** [reply]
Do you mean tag?

**Reply by Zstranslators** [reply]
Well, that's managing a cache of the same group or tag, and happy New Year

**Reply by hwbrzzl** [reply]
Thanks, happy New Year, too. We don't support tag for now, there is an issue to handle it: https://github.com/goravel/goravel/issues/174

**Comment by FengRuner** [comment]
I think the WithContext function in memory.go is error,because the Memory struct is not a new instance,I see redis struct ,that is ok

**Reply by hwbrzzl** [reply]
Actually, the ctx doesn't have any functions in the memory driver.

**Comment by haodiao** [comment]
建议redis驱动库 go-redis/redis/v8  升级为 redis/go-redis/v9 版本，理论上 go.mod 出现一个库的多个版本是不太好的

**Reply by hwbrzzl** [reply]
Yes, thanks for your feedback, will update it in v1.15.

**Reply by haodiao** [reply]
laravel 中有个隐性的调用，比如当默认缓存使用 redis 驱动时，使用 `Cache::connection()->hgetall("key")` 或指定驱动  `Cache::driver('redis')->connection()->hgetall("key")`就能直接使用 Cache 自带的 redis 连接配置来执行 redis 的操作命令，希望在 Goravel 也能增加上类似的调用方式，就能避免重复的写一遍连接。

**Reply by hwbrzzl** [reply]
Yes, please follow: https://github.com/goravel/goravel/issues/513

**Reply by haodiao** [reply]
`facades.Cache().Store()` 不是已经有了吗？这个 Cache::Store() 本质上还是框架缓存的操作逻辑和方法。  

但我说 laravel 的 Cache::connection() 是更底一层的连接，在缓存用 redis 时它就是一个 Redis::connection() 方法，在 Goravel 这里是指 go-redis 库的 `redis.NewClient()`，也就是复用缓存的连接方法去执行原始的 redis 命令，laravel Cache 的 redis 和 单独的 facades Redis 默认是共用一套连接方法的，`当然如果 Goravel 能增加一个专门 redis 的 facades 的话就更好了`。

**Reply by hwbrzzl** [reply]
Sorry, 发错了：https://github.com/goravel/goravel/issues/372

**Comment by jxgame** [comment]
按照 https://github.com/goravel/redis  这个进行配置Redis 一直不成功

当 CACHE_STORE=redis 启动即报错 
···
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x140 pc=0x423af10]
···

当 CACHE_STORE=custom 访问缓存时报错 
···
2024/11/26 13:11:05 http: panic serving 127.0.0.1:58486: runtime error: invalid memory address or nil pointer dereference
···

本机的Redis可以访问，Keys * 也正常

如何解决呢？

**Reply by jxgame** [reply]
但是在队列中使用正常 QUEUE_CONNECTION=redis  
Redis控制台 Keys *  时也能看到Task开头的

**Reply by hwbrzzl** [reply]
可以检查下顺序是否正确

<img width="677" alt="image" src="https://github.com/user-attachments/assets/eee725ee-5bba-4580-8a01-c6c36153c1c3">

**Reply by jxgame** [reply]
非常感谢，没仔细看这里的注释。的确是这个顺序的问题。感谢。

**Reply by jxgame** [reply]
这个地方，要将 &redis.ServiceProvider{}, 拿到 &cache.ServiceProvider{},（原位置不变） 的上面，如果将 &cache.ServiceProvider{} 拿到下面，编译时会报：
```
binding not found: goravel.cache
```
